optuna/distributions.py

```diff
@@ -678,3 +678,68 @@ def _get_single_value(distribution: BaseDistribution) -> Union[int, float, Categ
     elif isinstance(distribution, CategoricalDistribution):
         return distribution.choices[0]
     assert False
+
+
+# TODO(himkt): Remove this method with the deletion of deprecated distributions.
+# https://github.com/optuna/optuna/issues/2941
+def _convert_old_distribution_to_new_distribution(
+    distribution: BaseDistribution,
+    suppress_warning: bool = False,
+) -> BaseDistribution:
+
+    new_distribution: BaseDistribution
+
+    # Float distributions.
+    if isinstance(distribution, UniformDistribution):
+        new_distribution = FloatDistribution(
+            low=distribution.low,
+            high=distribution.high,
+            log=False,
+            step=None,
+        )
+    elif isinstance(distribution, LogUniformDistribution):
+        new_distribution = FloatDistribution(
+            low=distribution.low,
+            high=distribution.high,
+            log=True,
+            step=None,
+        )
+    elif isinstance(distribution, DiscreteUniformDistribution):
+        new_distribution = FloatDistribution(
+            low=distribution.low,
+            high=distribution.high,
+            log=False,
+            step=distribution.q,
+        )
+
+    # Integer distributions.
+    elif isinstance(distribution, IntUniformDistribution):
+        new_distribution = IntDistribution(
+            low=distribution.low,
+            high=distribution.high,
+            log=False,
+            step=distribution.step,
+        )
+    elif isinstance(distribution, IntLogUniformDistribution):
+        new_distribution = IntDistribution(
+            low=distribution.low,
+            high=distribution.high,
+            log=True,
+            step=distribution.step,
+        )
+
+    # Categorical distribution.
+    else:
+        new_distribution = distribution
+
+    if new_distribution != distribution and not suppress_warning:
+        message = (
+            f"{distribution} is deprecated and internally converted to"
+            f" {new_distribution}. If you see this message, please update your"
+            f" source code to use {new_distribution.__class__.__name__} instead of"
+            f" {distribution.__class__.__name__}."
+            " See https://github.com/optuna/optuna/issues/2941."
+        )
+        warnings.warn(message, FutureWarning)
```

Comment:
When I executed the following line, I got two deprecated warnings for a single line of code and I felt a bit noisy.

```python
trial = study.ask(fixed_distributions={f"x_{i}": optuna.distributions.UniformDistribution(0, 1) for i in range(n_params)})
```

 Do you think we can merge these warnings or remove the duplicated part in two messages?
```
./issue3420.py:23: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  trial = study.ask(fixed_distributions={f"x_{i}": optuna.distributions.UniformDistribution(0, 1) for i in range(n_params)})
/home/yanase/pfn/code/optuna-review/optuna/distributions.py:743: FutureWarning: UniformDistribution(high=1.0, low=0.0) is deprecated and internally converted to FloatDistribution(high=1.0, log=False, low=0.0, step=None). If you see this message, please update your source code to use FloatDistribution instead of UniformDistribution. See https://github.com/optuna/optuna/issues/2941.
```

Reply:
Thank you so much for the explanation. It totally makes sense!
I simplified the message in 211f90a.

Reply:
Thank you for your update. I confirmed that the redundant message was removed as follows: 

```
./issue3420.py:23: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  trial = study.ask(fixed_distributions={f"x_{i}": optuna.distributions.UniformDistribution(0, 1) for i in range(n_params)})
/home/yanase/pfn/code/optuna-review/optuna/distributions.py:740: FutureWarning: UniformDistribution(high=1.0, low=0.0) is deprecated and internally converted to FloatDistribution(high=1.0, log=False, low=0.0, step=None). See https://github.com/optuna/optuna/issues/2941.
  warnings.warn(message, FutureWarning)
```

BTW, the rest formatted text can be seen in 
> :class:`~optuna.distributions.FloatDistribution`

but it can be removed simply because it is not rendered in the console. I don't think this is in the scope of this PR.